lib/models/package-info-cache/package-info.js

```diff
@@ -415,6 +490,126 @@ class PackageInfo {
+   * calls to the constructor! A given addon is either a singleton or not for an
+   * entire run.
+   *
+   * @method getAddonInstance
+   * @param {} parent the addon/project that is to be the direct parent of the
+   * addon instance created here
+   * @param {*} project the project that is to contain this addon instance
+   * @return {Object} the constructed instance of the addon
+   */
+  getAddonInstance(parent, project) {
+    let ctor = this.getAddonConstructor();
+    let addon;
+
+    if (!isSupportingSingletons || !ctor.isSingleton) {
+      // We have a 'regular' addon or we're not supporting singleton addons at all,
+      // so just return a regular instance. It is not in any singleton cache.
+      addon = this.constructAddonInstance(parent, project);
+
+      if (addon.initializeAddons) {
+        addon.initializeAddons();
+      } else {
+        addon.addons = [];
+      }
+
+      return addon;
+    }
+
+    // Find the bundleRoot that will contain this singleton instance.
+    // Find some parent, if there is one, that returns true for PerBundleAddonCache.isBundleRoot(obj).
+    // 'Project' is always the ultimate bundle.
+
+    // An optimization that lazy engines does that we can do as well. If the bundle
+    // root is not the project, but the project does or is going to have an instance
+    // of this singleton in its own cache, then use the project cache.
+    let bundleRoot = this.findBundleRoot(parent);
+    let bundledProjectAddons = project.getBundledAddonPackageInfos();
+
+    if (bundleRoot === project || (bundleRoot !== project && !bundledProjectAddons.has(this))) {
+      // the instance belongs to its bundle root, whether it's the project or not.
+
+      // Luckily for us, both types of bundle root (Project and lazy engine) have a 
+      // _packageInfo field, just what we need!
+      let bundleRootPkgInfo = bundleRoot._packageInfo;
+
+      let cacheEntry = project.perBundleAddonCache.getPerBundleCacheEntry(bundleRootPkgInfo, this);
+
+      if (cacheEntry && cacheEntry.isBuilt) {
+        // Return a proxy for the already-set-up cache entry.
+        console.log(`BR ${bundleRootPkgInfo.realPath}: PROXY: ${this.realPath}`);
+        let addonProxy = getAddonProxy(cacheEntry, parent);
+        return addonProxy;
+      } else {
+        // The cache entry may or may not exist, but the instance isn't set up yet.
+        // Create a new instance, install it in the bundleRoot's cache (add/update
+        // the cache entry) and return the new instance.
+        console.log(`BR ${bundleRootPkgInfo.realPath}: NEWENTRY: ${this.realPath}`);
+        addon = this.constructAddonInstance(parent, project);
+        project.perBundleAddonCache.addPerBundleInstance(bundleRootPkgInfo, this, addon);
+
+        if (addon.initializeAddons) {
+          addon.initializeAddons();  // XXX based on addon.setupRegistry, I think this is basically a no-op by this point.
```

Comment:
We still have to call it (because addon's could decide to have custom constructors and not have the `setupRegistry` in constructor issue). Should tweak the comment but keep the actual implementation.

Reply:
Moved the block of 4 lines starting at 595 to a separate function, which is then invoked (and which is also invoked in the PerBundleAddonCache), and put a more enlightening comment.